src/plugins/presentation_util/public/components/floating_actions/floating_actions.tsx

```diff
@@ -28,45 +48,56 @@ export const FloatingActions: FC<FloatingActionsProps> = ({
   const { euiTheme } = useEuiTheme();
   const anchorRef = useRef<HTMLSpanElement>(null);
   const actionsRef = useRef<HTMLDivElement>(null);
+
+  const [position, setPosition] = useState<FloatingActionsPosition>({
+    top: 0,
+    left: 0,
+  });
   const [areFloatingActionsVisible, setFloatingActionsVisible] = useState<boolean>(false);
 
   const showFloatingActions = useCallback(
-    () => isEnabled && setFloatingActionsVisible(true),
-    [isEnabled, setFloatingActionsVisible]
+    () => isEnabled && !areFloatingActionsVisible && setFloatingActionsVisible(true),
+    [isEnabled, areFloatingActionsVisible, setFloatingActionsVisible]
   );
   const hideFloatingActions = useCallback(
-    () => setFloatingActionsVisible(false),
-    [setFloatingActionsVisible]
+    () => areFloatingActionsVisible && setFloatingActionsVisible(false),
+    [areFloatingActionsVisible, setFloatingActionsVisible]
   );
 
-  const anchorBoundingRect = anchorRef.current?.getBoundingClientRect();
-  const actionsBoundingRect = actionsRef.current?.getBoundingClientRect();
+  const updatePosition = useCallback(() => {
+    if (anchorRef.current && actionsRef.current) {
+      const anchorBoundingRect = anchorRef.current?.getBoundingClientRect();
+      const actionsBoundingRect = actionsRef.current?.getBoundingClientRect();
 
-  const hiddenActionsStyles = `
-  visibility: hidden;
-  opacity: 0;
+      const top =
+        anchorBoundingRect.top -
+        (usingTwoLineLayout ? parseInt(euiTheme.size.xs, 10) : parseInt(euiTheme.size.l, 10)) +
+        window.scrollY;
+      const left =
+        anchorBoundingRect.right -
+        actionsBoundingRect.width -
+        parseInt(euiTheme.size.xs, 10) +
+        window.scrollX;
 
-  // slower transition on hover leave in case the user accidentally stops hover
-  transition: visibility 0.3s, opacity 0.3s;
-  `;
-  const visibleActionsStyles = `
-  transition: visibility 0.1s, opacity 0.1s;
-  visibility: visible;
-  opacity: 1;
+      if (position.top !== top || position.left !== left) {
+        setPosition({ top, left });
+      }
+    }
+  }, [anchorRef, actionsRef, euiTheme.size, position, usingTwoLineLayout]);
+
+  const floatingActionsStyles = css`
+    top: ${position.top}px;
+    left: ${position.left}px;
+
+    ${areFloatingActionsVisible ? visibleActionsStyles : hiddenActionsStyles}
   `;
 
-  const floatingActionStyles =
-    anchorBoundingRect && actionsBoundingRect
-      ? css`
-          top: ${anchorBoundingRect.top -
-          (usingTwoLineLayout ? parseInt(euiTheme.size.xs, 10) : parseInt(euiTheme.size.l, 10))}px;
-          left: ${anchorBoundingRect.right -
-          actionsBoundingRect.width -
-          parseInt(euiTheme.size.xs, 10)}px;
+  useEffect(updatePosition);
```

Comment:
Yeah, I added it because it wasn't correctly calculating the initial position of the floating actions, but you're right that we don't want all these extra re-renders especially when it's out of view. The issue with calling it once was I needed to somehow call it after the DOM elements have already rendered so I can use the refs to calculate the position correctly which wasn't working when I called `updatePosition` in a `useMount`. After digging into it some more, I found out if I used a `useLayoutEffect` instead, it'll wait for all the DOM changes before firing, and it seems to be doing the trick.

Reply:
That's a great solution! `useLayoutEffect` is new to me - but that's super cool 🎉